src/libmongoc/src/mongoc/mongoc-crypt.c

```diff
@@ -597,23 +614,35 @@ _state_need_kms (_state_machine_t *state_machine, bson_error_t *error)
          uint32_t bytes_needed = mongocrypt_kms_ctx_bytes_needed (kms_ctx);
          ssize_t read_ret;
 
+         sleep_usec = mongocrypt_kms_ctx_usleep (kms_ctx);
+         if (sleep_usec > 0) {
+            _mongoc_usleep (sleep_usec);
+         }
+
          /* Cap the bytes requested at the buffer size. */
          if (bytes_needed > BUFFER_SIZE) {
             bytes_needed = BUFFER_SIZE;
          }
 
          read_ret = mongoc_stream_read (tls_stream, buf, bytes_needed, 1 /* min_bytes. */, sockettimeout);
-         if (read_ret == -1) {
-            bson_set_error (
-               error, MONGOC_ERROR_STREAM, MONGOC_ERROR_STREAM_SOCKET, "failed to read from KMS stream: %d", errno);
-            goto fail;
+         if (read_ret <= 0) {
+            if (mongocrypt_kms_ctx_fail (kms_ctx)) {
+               continue;
+            } else {
+               if (read_ret == -1) {
+                  bson_set_error (error,
```

Comment:
When `mongocrypt_kms_ctx_fail` returns false, I suggest also including the error from the `kms_ctx` (which may indicate the max retries was hit):

```c
bson_error_t kms_error;
BSON_ASSERT (!_kms_ctx_check_error (kms_ctx, &kms_error, true));
bson_set_error (error,
                MONGOC_ERROR_STREAM,
                MONGOC_ERROR_STREAM_SOCKET,
                "%s. Failed to read from KMS stream to: %s",
                kms_error.message,
                endpoint);
goto fail;
```

Testing with the mock server (by setting `count` to 5) changed the error from:

```
failed to read from KMS stream: 0
```

To:
```
KMS request failed after 3 retries due to a network error. Failed to read from KMS stream to: 127.0.0.1:9003
```

I am not sure why `errno` is 0 in that case (possible separate bug with `mongoc_stream_read`?). But I think that is out of scope for this PR.

Suggestion also applies to other `mongocrypt_kms_ctx_fail` checks.